File: ctci/ch2/LinkedList.cpp
#include <iostream>

using namespace std;

class Node
{
public:
    Node *next;
    int data;

    Node(int value) : data(value), next(nullptr) {}
    Node(Node *head, int value) : data(value), next(head) {}
};

class LinkedList
{
public:
    Node *head;
    Node *tail;

    LinkedList() : head(nullptr) {}
    LinkedList(Node *node) : head(node) {}

    void insertHead(int value)
    {
        Node *newNode = new Node(value);
        newNode->next = head;
        head = newNode;
    }

    void insertTail(int value)
    {
        Node *newNode = new Node(value);
        if (head == nullptr)
        {
            head = newNode;
            return;
        }
        Node *curr = head;
        while (curr->next != nullptr)
        {
            curr = curr->next;
        }
        curr->next = newNode;
    }

    void insertTail(Node* newNode)
    {
        if (head == nullptr)
        {
            head = newNode;
            return;
        }
        Node *curr = head;
        while (curr->next != nullptr)
        {
            curr = curr->next;
        }
        curr->next = newNode;
    }

    void deleteList()
    {
        Node *nextNode;
        while (head)
        {
            nextNode = head->next;
            delete (head);
            head = nextNode;
        }
    }

    void printList()
    {
        Node *cur = head;
        while (cur)
        {
            cout << cur->data << " ";
            cur = cur->next;
        }
        cout << endl;
    }
};

File: ctci/ch2/LinkedList.py
class Node:
    def __init__(self, data, next=None, prev=None):
        self.data = data
        self.next = next
        self.prev = prev

    def __str__(self):
        return str(self.data)


class LinkedList:
    def __init__(self, datas=None):
        self.head = None
        self.tail = None
        if datas is not None:
            for data in datas:
                self.insert(data)

    def __iter__(self):
        curr = self.head
        while curr:
            yield curr
            curr = curr.next

    def __str__(self):
        values = [str(x) for x in self]
        return " ".join(values)

    def insert(self, data):
        if self.head is None:
            self.head = Node(data)
            self.tail = self.head
        else:
            self.tail.next = Node(data)
            self.tail = self.tail.next
